### PR TITLE
[Artifacts] Fix resolving target path in artifacts html

### PR DESCRIPTION
**Pull request by alonmr**
When we show the results html in a jupyter notebook we link the artifacts to their target paths.
If logging the artifact failed and `generate_target_path_from_artifact_hash` is enabled the target path might not have been set - if that happens, warn and continue.

https://jira.iguazeng.com/browse/ML-2880

### CODE REVIEW

File: mlrun/render.py [hunk 1]
```diff
@@ -96,20 +99,33 @@ def link_html(text, link=""):
 
 
 def artifacts_html(x, pathcol="path"):
-    if not x:
+    artifact_dicts, output_path = x
```
Comment (TomerShor) [inline mlrun/render.py:102]: Can we give `x` a more informative name?
I know it's called from the lambda function but still

File: mlrun/render.py [hunk 2]
```diff
@@ -96,20 +99,33 @@ def link_html(text, link=""):
...
+            pathcol == "target_path"
+            and config.artifacts.generate_target_path_from_artifact_hash
+        ):
+            artifact = mlrun.artifacts.dict_to_artifact(artifact_dict)
+            target_path = artifact.resolve_target_path(artifact_path=output_path)
+            link, ref = link_to_ipython(target_path)
+
+        else:
+            # TODO: remove this in 1.5.0 once we no longer support legacy format
+            if pathcol in artifact_dict:
```
Comment (TomerShor) [inline mlrun/render.py:120]: you can also use the function `is_legacy_artifact`, since this is what the condition means.

File: mlrun/render.py [hunk 3]
```diff
@@ -96,20 +99,33 @@ def link_html(text, link=""):
...
+                link, ref = link_to_ipython(artifact_dict[pathcol])
+            else:
+                link, ref = link_to_ipython(artifact_dict["spec"][pathcol])
+
+        if "key" in artifact_dict:
```
Comment (TomerShor) [inline mlrun/render.py:125]: same here

**Review by TomerShor — COMMENTED** [review]
Minor comments and suggestions

File: mlrun/render.py (hunk 1 above)
Comment (Tankilevitch) [inline mlrun/render.py:102]: we gotta have more description in this area, most of it wasn't been touched from 2019, please try to add more information

Comment (Tankilevitch) [reply mlrun/render.py:102]: also type hinting would be useful, as well as changing the `pathcol` maybe to a more indicative name such as `attribute_name` ( just an example open for other naming )

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
added some comments, 
Also I am not sure when is the exact case that target path isn't being enriched, because it does been enriched when logging dataset, model etc.. , please elaborate more on the exact case

File: mlrun/render.py (hunk 1 above)
Comment (alonmr) [reply mlrun/render.py:102]: 👍

File: mlrun/render.py [hunk 4]
```diff
@@ -370,12 +405,15 @@ def time_str(x):
...
     else:
         df["labels"] = df["labels"].apply(dict_html)
         df["inputs"] = df["inputs"].apply(inputs_html)
-        df["artifacts"] = df["artifacts"].apply(
-            lambda x: artifacts_html(x, "target_path")
+        df["artifacts"] = df[["artifacts", "output_path"]].apply(
+            lambda artifacts_tuple: artifacts_html(artifacts_tuple, "target_path"),
+            axis=1,
         )
+        df.drop("output_path", axis=1, inplace=True)
```
Comment (Tankilevitch) [inline mlrun/render.py:416]: why

Comment (alonmr) [reply mlrun/render.py:416]: because I don't want it to be part of the html that is presented in the jupy (it is being displayed late in get_tblframe)

Comment (alonmr) [reply mlrun/render.py:420]: added a comment

File: tests/test_render.py [hunk 5]
```diff
@@ -12,32 +12,58 @@
...
+    "generate_artifact_hash_mode, expected_target_paths",
+    [
+        # (
+        #     False,
+        #     [
+        #         f"{results}/log-function-log-dataset/0/feature_1.csv",
+        #         f"{results}/log-function-log-dataset/0/feature_2.csv",
+        #     ],
+        # ),
+        (
```
Comment (Tankilevitch) [inline tests/test_render.py:41]: need to uncomment

File: mlrun/artifacts/base.py [hunk 6]
```diff
@@ -381,6 +381,42 @@ def _upload_file(
...
+        """
+        Resolve the target path for the artifact,
+        if the user didn't specify a target path, and mlrun.mlconf.artifacts.generate_target_path_from_artifact_hash is
+        True, the target path will be resolved from the artifact hash
+        if the artifact has a body, the target path will be resolved from
+        the body, otherwise it will be resolved from the source path
+
+        :param artifact_path:   the artifact path to use when generating the target path
+        :return:                the target path
+        """
```
Comment (Tankilevitch) [inline mlrun/artifacts/base.py:394]: redundant, this doesn't seem to be used anymore

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

**Review by Tankilevitch — APPROVED** [review]
LGTM

**Pull request merged**